Feature: interface query transformation

edarroudi opened the pull request:
Hi guys,

I created a PR for how to implement the QueryTransformation interface as an extension. See #1947 

Added documentation now, tested as well.

Still: **An important questsion for me is whether to run the query transform before running resolveTokens or as it is like now, afterwards.
![image](https://user-images.githubusercontent.com/935493/165065808-edd2f967-98ad-4ba4-9031-6a3044117b74.png)**

As for the default query transformer: There doesn't have to be one, i just thought it could be useful to have one. I'm open to any suggestions here, maybe a possibility for users to add custom regex.

![image](https://user-images.githubusercontent.com/935493/170033948-8b2be700-d116-4373-847a-08509f80d99b.png)

![image](https://user-images.githubusercontent.com/935493/170034743-2fd38e0a-a710-490e-b1c2-632401cd2462.png)

wobba commented:
I think running it POST token substitution is the correct thing to do. Because then you replace tokens with potential terms, which then should be transformed.

wobba commented on search-parts/src/dataSources/MicrosoftSearchDataSource.ts:
```diff
@@ -595,6 +595,21 @@ export class MicrosoftSearchDataSource extends BaseDataSource<IMicrosoftSearchDa
+        const clonedContext = cloneDeep(dataContext);
+         
+        for (const modifier of selectedCustomQueryModifier) {
+             const resp = await modifier.modifyQuery({queryTemplate:cloneDeep(queryTemplate), queryText: cloneDeep(queryText)}, clonedContext, this._tokenService.resolveTokens);            
```
Do you need cloneDeep on queryText which is a string?

edarroudi replied:
no thanks ;-) came from refactoring :(

wobba commented on search-parts/src/dataSources/SharePointSearchDataSource.ts:
```diff
@@ -768,6 +768,19 @@ export class SharePointSearchDataSource extends BaseDataSource<ISharePointSearch
+        const clonedContext = cloneDeep(dataContext);
+        for (const modifier of selectedCustomQueryModifier) {
+        
+            const resp = await modifier.modifyQuery({queryTemplate:cloneDeep(searchQuery.QueryTemplate), queryText: cloneDeep(searchQuery.Querytext)}, clonedContext, this._tokenService.resolveTokens);            
```
Same here, is cloneDeep on querytext needed? And do we have a casing issue as it's `Querytext`. Should be `searchQuery.queryText` or `searchQuery.QueryText`?

edarroudi replied:
The casing issue isn't from my code. I can and will fix it in another branch, but I would have to check for any side effects (probaly 0 % chance for any sideffects - but you never know...)

wobba replied:
You can leave it as is for now.. if it impacts an interface we might want to leave it.

edarroudi commented:
> I think running it POST token substitution is the correct thing to do. Because then you replace tokens with potential terms, which then should be transformed.

It probably doesn't matter anyways because it could be possible to use the resolveToken / context to generate the whole thing again in the custom query function.

wobba commented:
I believe I closed this by accident.

edarroudi commented:
I'll do a new test / fix issues now => converting it to draft again

edarroudi commented:
Hey @wobba , I did some tests now and it works fine. I updated the integrated "Word Modifier" to check for KQL keywords. Don't know if this will work without issues. I would invest more time in that but I don't know if you guys want to have the internal "Word Modifier" or should I export that to a sample extension? Or maybe you guys would like another Internal Modifier, maybe something like a REGEX modifier? Would appreciate any answer so I can finish this up. Thanks!

wobba commented:
Can you please split the extensibility interface changes into a separate PR, as that has to be merged, publish a new npm packages, which can then be used in the web parts.

FranckyC commented:
@wobba can you wait before merging ? I would like to review it before. Thanks!

edarroudi commented:
> Can you please split the extensibility interface changes into a separate PR, as that has to be merged, publish a new npm packages, which can then be used in the web parts.

did as requested. Still unsure about the base WordModifier and it's value to be available inside the main webpart package...

FranckyC commented:
> > Can you please split the extensibility interface changes into a separate PR, as that has to be merged, publish a new npm packages, which can then be used in the web parts.
> 
> did as requested. Still unsure about the base WordModifier and it's value to be available inside the main webpart package...

See my review @edarroudi. This logic should be moved outside of the main WebPart and implemented as property pane options for a particular modifier.

edarroudi commented:
> > > Can you please split the extensibility interface changes into a separate PR, as that has to be merged, publish a new npm packages, which can then be used in the web parts.
> > 
> > 
> > did as requested. Still unsure about the base WordModifier and it's value to be available inside the main webpart package...
> 
> See my review @edarroudi. This logic should be moved outside of the main WebPart and implemented as property pane options for a particular modifier.

@FranckyC  I don't see any review. Maybe I'm missing something? As far as i understand you, I'll extract the WordModifier and add it later as a sample for the extensions here https://github.com/microsoft-search/pnp-modern-search-extensibility-samples/tree/main/search-extensibility-samples/src/libraries ?

FranckyC commented:
@edarroudi my mistake I forgot to publish.

edarroudi commented on search-parts/src/dataSources/MicrosoftSearchDataSource.ts:
```diff
@@ -215,15 +215,15 @@ export class MicrosoftSearchDataSource extends BaseDataSource<IMicrosoftSearchDa
     }
 
-    public async getData(dataContext: IDataContext): Promise<IMicrosoftSearchDataSourceData> {
+    public async getData(dataContext: IDataContext, selectedCustomQueryModifier?: IQueryModifier[]): Promise<IMicrosoftSearchDataSourceData> {
```
But that would mean that we can only change the queryText and not the queryTemplate as that is datasource specific.

FranckyC replied:
Correct, query template is a SharePoint/Microsoft Search concept. If in the future we add a new builtin data source or even custom ones with extensibility, they won't have this option. This has to be the most generic as possible, not shortcuts :)

edarroudi replied:
ok will update the interface in https://github.com/microsoft-search/pnp-modern-search/pull/2450 very soon and then push the fixed version of this PR

edarroudi commented:
I did an update now and tested it with a custom extension. Question is at one point if I can optimize setTokens using the calculated inputQueryText once?

edarroudi commented:
Ready for review. Tested it multiple times with different configurations.

FranckyC commented:
@edarroudi I don't see where the modifier logic is actually implemented (basically the `modifyQuery` method) :/ I would also expect a PR for a sample query modifier here https://github.com/microsoft-search/pnp-modern-search-extensibility-samples/pulls working with this version.

edarroudi commented on search-parts/src/webparts/searchResults/SearchResultsWebPart.ts:
```diff
@@ -2417,17 +2515,37 @@ export default class SearchResultsWebPart extends BaseWebPart<ISearchResultsWebP
+        for (const modifier of this._selectedCustomQueryModifier) {
+
+            //Cloned context won't be correct for inputQueryText after first modification!
+            const modifiedQueryText = await modifier.modifyQuery(queryText, clonedContext, this.tokenService.resolveTokens.bind(this.tokenService));
```
@FranckyC  This is where the modifyQuery method is called.

edarroudi commented:
> @edarroudi I don't see where the modifier logic is actually implemented (basically the `modifyQuery` method) :/ I would also expect a PR for a sample query modifier here https://github.com/microsoft-search/pnp-modern-search-extensibility-samples/pulls working with this version.

Hey @FranckyC  - I'll merge the interface into this PR and add a PR for the extensibility samples tonight (already prepared).
I commented the line where the modifyQuery method is called.

edarroudi commented:
Hi @FranckyC, as requested I added the interface to this branch and added a sample here https://github.com/microsoft-search/pnp-modern-search-extensibility-samples/pull/13

edarroudi commented:
@wobba 

I changed the eslint rule from 2600 to 3000 
'max-lines': ['warn', { max: 3000 }],

Do you guys have already a plan to refactor the whole SearchResultsWebpart? I find it hard to work with that huge file (actually the methods are way too big for me as well, containing like 100 lines of code).

Can I be of any support for that? For a custom solution I started to extract the PropertyPane code in other classes  so that at least that code is not "disturbing the flow"...

wobba commented:
@edarroudi I'm all for refactoring it - and you can PR that separately if you want.

FranckyC commented on search-extensibility/src/models/queryModifier/BaseQueryModifier.ts:
```diff
@@ -0,0 +1,58 @@
+        // Do nothing by default      
+    }
+
+    public modifyQuery(queryText: string, dataContext: IDataContext, resolveTokens: (string: string) => Promise<string>): Promise<string> {
```
I don't see the point to pass the `dataContext` and `resolveTokens` method here. This is not the responsibility of the modifier to resolve tokens. I don't see a scenario where tokens resolved by a modifier will be different that the main Web Part. In your sample, you don't use these parameters.

The `modifyQuery` operation should occur before the tokens resolved representing the "raw" keywords without any modifications. Also because the modifier concerns only the "queryText" parameter and not query template, I don't expect many tokens here.

edarroudi replied:
done

FranckyC commented on search-parts/src/webparts/searchResults/SearchResultsWebPart.ts:
```diff
@@ -617,6 +653,27 @@ export default class SearchResultsWebPart extends BaseWebPart<ISearchResultsWebP
+                    });
+                }
+
+                propertyPanePages.push({
```
It is odd to have a dedicated configuration page added just for query modifiers. Because it only influence the input query text settings, I would expect the configuration to be placed here:
![image](https://user-images.githubusercontent.com/5659122/201162155-949c8be3-ca0b-4593-90ec-fd5cae96d32d.png)

edarroudi replied:
done

FranckyC commented on search-parts/src/webparts/searchResults/SearchResultsWebPart.ts:
```diff
@@ -2526,4 +2644,176 @@ export default class SearchResultsWebPart extends BaseWebPart<ISearchResultsWebP
+    private _getQueryModifierFields(): IPropertyPaneField<any>[] {
+
+        let queryTransformationFields: IPropertyPaneField<any>[] = [
+            PropertyPaneToggle("enableCustomQueryTransformation", {
```
I would simply remove this field as it is not necessary. In your implementation, there are 3 places where modifiers can be enable/disabled:
- Enable/Disable in the extensibility library settings (with other extensions as well)
- Enable/Disable all modifiers at once using the toggle
- Enable/Disabled a query modifier individually from configuration page.

If the user doesn't want to have any modifiers applied, he can just disable them one by one, the same way as libraries, no need to have a dedicated toggle.

edarroudi replied:
done

FranckyC commented:
@edarroudi keep the property pane fields but add them in the connection page then, below the connections. Not optimal in terms of UI but I can live with it. Of course, if no fields are provided, no section should appear.

edarroudi commented on search-parts/src/webparts/searchResults/SearchResultsWebPart.ts:
```diff
@@ -240,6 +267,16 @@ export default class SearchResultsWebPart extends BaseWebPart<ISearchResultsWebP
+
+            //Check if any custom query modifier is active and modify the inputQueryText
+            if (this._selectedCustomQueryModifier && this._selectedCustomQueryModifier.length > 0) {
+                this._currentDataContext.inputQueryText = await this.getModifiedInputQueryText(this._currentDataContext.inputQueryText, this._currentDataContext);
```
done

edarroudi commented on search-parts/src/webparts/searchResults/SearchResultsWebPart.ts:
```diff
@@ -2526,4 +2644,176 @@ export default class SearchResultsWebPart extends BaseWebPart<ISearchResultsWebP
+        }
+    }
+
+    private _getQueryModifierFields(): IPropertyPaneField<any>[] {
```
as discussed property pane fields stay and are moved to the correct position @ connections

edarroudi commented:
Implemented the requested changes & tested. Also updated the sample PR : https://github.com/microsoft-search/pnp-modern-search-extensibility-samples/pull/13

Thank you

FranckyC commented:
Hi @edarroudi, approved and merged. Thanks again for your hard work, this new feature is really cool! (and sorry for the delay in the review).
